config/remote_config.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	return body, nil
+}
+
+func (c *RemoteConfig) Reset(targetPath string) (FileRemovalStatus, error) {
+	fileInfo, err := os.Lstat(targetPath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return FileNotFound, nil
+		} else if errors.Is(err, os.ErrPermission) {
+			return PermissionDenied, fmt.Errorf("permission denied: %w", err)
+		}
+		return UnknownError, fmt.Errorf("failed to stat file: %w", err)
+	}
+
+	if fileInfo.IsDir() {
+		return FileIsDirectory, fmt.Errorf("cannot remove directory: %w", err)
+	}
+
+	if err := os.Remove(targetPath); err != nil {
+		if errors.Is(err, os.ErrPermission) {
+			return PermissionDenied, fmt.Errorf("permission denied: cannot remove %w", err)
+		}
+		return UnknownError, fmt.Errorf("error removing file: %w", err)
+	}
+
+	return FileRemoved, nil
+}
+
+func retry[T any](ctx context.Context, maxAttempts int, fn func(ctx context.Context) (T, error)) (T, error) {
+	var result T
+	var err error
+
+	initialDelay := 1 * time.Second
+	maxDelay := 30 * time.Second // Added maxDelay
+
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		result, err = fn(ctx)
+
+		if err == nil {
+			return result, nil
+		}
+
+		// Calculate exponential backoff with jitter.
+		delay := initialDelay * time.Duration(math.Pow(2, float64(attempt)))
+		// Add some jitter (using crypto/rand for security).
+		jitter, randErr := rand.Int(rand.Reader, big.NewInt(int64(delay/4)))
+		if randErr != nil {
+			// Fallback to a less secure random number if crypto/rand fails.
+			jitter = big.NewInt(m.Int63n(int64(delay / 4)))
+		}
+		delay += time.Duration(jitter.Int64())
+
+		if delay > maxDelay {
+			delay = maxDelay
+		}
+
+		select {
+		case <-time.After(delay): // Wait for the delay.
+		case <-ctx.Done(): // Check for context cancellation.
+			return result, ctx.Err() // Return context error if cancelled.
+		}
+	}
+
+	return result, fmt.Errorf("failed after %d attempts: %w", maxAttempts, err) // Wrap the final error.
+}
+
+func writeToDisk(content []byte, filePath string) (int, error) {
+	// create directory path if not already created
+	dirPath := filepath.Dir(filePath)
+	if err := os.MkdirAll(dirPath, 0700); err != nil {
+		return 0, fmt.Errorf("failed to create directory ~/.config/gitleaks : %w", err)
+	}
+
+	// Create a temporary file in the same directory
+	tempFile, err := os.CreateTemp(dirPath, "gitleaks-config-*.tmp")
+	if err != nil {
+		return 0, fmt.Errorf("failed to create temporary file: %w", err)
+	}
+	tempFilePath := tempFile.Name()
+	defer func() {
+		if removeErr := os.Remove(tempFilePath); removeErr != nil {
+			err = fmt.Errorf("failed to remove temporary file: %w (original error: %w)", removeErr, err)
+		} // Clean up the temp file in case of failure
+	}()
+
+	// Write to the temporary file
+	writer := bufio.NewWriter(tempFile)
+	bytesWritten, err := writer.Write(content)
+	if err != nil {
+		return 0, fmt.Errorf("failed to write to temporary file: %w", err)
+	}
```

Comment:
nit: if you pass in the content as a io.Reader this could be a call to `io.Copy`. Only downside is you wouldn't get the empty file check.